presto-main/src/main/java/com/facebook/presto/operator/AggregationOperator.java

```diff
@@ -133,115 +200,440 @@ public boolean isFinished()
+            BOOLEAN.writeBoolean(falseBlockBuilder, false);
+            FALSE_SINGLE_VALUE_BLOCK = falseBlockBuilder.build();
+        }
+
+        private final int intermediateIsDistinctCount;
+        private final List<Boolean> intermediateIsDistinct;
+        private final List<Type> intermediateTypes;
+        private final List<AggregationInputChannel> inputs;
+        private final List<Block> inputSingleNullBlocks;
+
+        public IntermediateLayout(List<InternalAccumulatorFactory> accumulatorFactories, List<AggregationInputChannel> inputs)
+        {
+            requireNonNull(accumulatorFactories, "accumulatorFactories is null");
+            requireNonNull(inputs, "inputs is null");
+            this.intermediateIsDistinctCount = (int) accumulatorFactories.stream()
+                    .filter(InternalAccumulatorFactory::isDistinct)
+                    .count();
+            this.intermediateIsDistinct = accumulatorFactories.stream()
+                    .map(InternalAccumulatorFactory::isDistinct)
+                    .collect(toImmutableList());
+            this.intermediateTypes = accumulatorFactories.stream()
+                    .map(InternalAccumulatorFactory::getIntermediateType)
+                    .collect(toImmutableList());
+
+            this.inputs = ImmutableList.copyOf(requireNonNull(inputs, "inputs is null"));
+            this.inputSingleNullBlocks = accumulatorFactories.stream()
+                    .map(accumulatorFactory -> accumulatorFactory.getIntermediateType().createBlockBuilder(null, 1)
+                            .appendNull()
+                            .build())
+                    .collect(toImmutableList());
+        }
+
+        private int getOutputChannelCount()
+        {
+            return 1 + intermediateTypes.size() + inputs.size();
+        }
+
+        public boolean hasDistinct()
+        {
+            return intermediateIsDistinctCount > 0;
+        }
+
+        public boolean hasNonDistinct()
+        {
+            return intermediateIsDistinctCount < intermediateIsDistinct.size();
         }
 
-        public void processPage(Page page)
+        private int getIntermediateChannel(int accumulatorIndex)
         {
-            if (step.isInputRaw()) {
-                aggregation.addInput(page);
+            checkElementIndex(accumulatorIndex, intermediateTypes.size());
+            return 1 + accumulatorIndex;
+        }
+
+        private Optional<Page> buildPartialDistinctRows(Page page, List<Optional<Block>> partialDistinctMasks)
+        {
+            checkArgument(partialDistinctMasks.size() == intermediateTypes.size());
+            verify(partialDistinctMasks.stream().filter(Optional::isPresent).allMatch(mask -> mask.get().getPositionCount() == page.getPositionCount()));
+
+            // check if any positions are distinct, this shouldn't happen since functions shouldn't return an empty mask
```

Comment:
my reading of this comment is that it shouldn't happen that any position is distinct, which is probably not what it is intended